--- orig/src/date.c	2014-08-27 12:25:53.000000000 -0400
+++ new/src/date.c	2014-08-27 14:04:01.000000000 -0400
@@ -4,7 +4,19 @@
  * Copyright (C) Linus Torvalds, 2005
  */
 
-#include "cache.h"
+#include <sys/time.h>
+#include <ctype.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <limits.h>
+#include <stdarg.h>
+
+#include "date.h"
+
+#include <R.h>
+
+#define ARRAY_SIZE(x) (sizeof(x)/sizeof(x[0]))
 
 /*
  * This is like mktime, but without normalization of tm_wday and tm_yday.
@@ -86,137 +98,6 @@
 	return offset * eastwest;
 }
 
-void show_date_relative(unsigned long time, int tz,
-			       const struct timeval *now,
-			       struct strbuf *timebuf)
-{
-	unsigned long diff;
-	if (now->tv_sec < time) {
-		strbuf_addstr(timebuf, _("in the future"));
-		return;
-	}
-	diff = now->tv_sec - time;
-	if (diff < 90) {
-		strbuf_addf(timebuf,
-			 Q_("%lu second ago", "%lu seconds ago", diff), diff);
-		return;
-	}
-	/* Turn it into minutes */
-	diff = (diff + 30) / 60;
-	if (diff < 90) {
-		strbuf_addf(timebuf,
-			 Q_("%lu minute ago", "%lu minutes ago", diff), diff);
-		return;
-	}
-	/* Turn it into hours */
-	diff = (diff + 30) / 60;
-	if (diff < 36) {
-		strbuf_addf(timebuf,
-			 Q_("%lu hour ago", "%lu hours ago", diff), diff);
-		return;
-	}
-	/* We deal with number of days from here on */
-	diff = (diff + 12) / 24;
-	if (diff < 14) {
-		strbuf_addf(timebuf,
-			 Q_("%lu day ago", "%lu days ago", diff), diff);
-		return;
-	}
-	/* Say weeks for the past 10 weeks or so */
-	if (diff < 70) {
-		strbuf_addf(timebuf,
-			 Q_("%lu week ago", "%lu weeks ago", (diff + 3) / 7),
-			 (diff + 3) / 7);
-		return;
-	}
-	/* Say months for the past 12 months or so */
-	if (diff < 365) {
-		strbuf_addf(timebuf,
-			 Q_("%lu month ago", "%lu months ago", (diff + 15) / 30),
-			 (diff + 15) / 30);
-		return;
-	}
-	/* Give years and months for 5 years or so */
-	if (diff < 1825) {
-		unsigned long totalmonths = (diff * 12 * 2 + 365) / (365 * 2);
-		unsigned long years = totalmonths / 12;
-		unsigned long months = totalmonths % 12;
-		if (months) {
-			struct strbuf sb = STRBUF_INIT;
-			strbuf_addf(&sb, Q_("%lu year", "%lu years", years), years);
-			strbuf_addf(timebuf,
-				 /* TRANSLATORS: "%s" is "<n> years" */
-				 Q_("%s, %lu month ago", "%s, %lu months ago", months),
-				 sb.buf, months);
-			strbuf_release(&sb);
-		} else
-			strbuf_addf(timebuf,
-				 Q_("%lu year ago", "%lu years ago", years), years);
-		return;
-	}
-	/* Otherwise, just years. Centuries is probably overkill. */
-	strbuf_addf(timebuf,
-		 Q_("%lu year ago", "%lu years ago", (diff + 183) / 365),
-		 (diff + 183) / 365);
-}
-
-const char *show_date(unsigned long time, int tz, enum date_mode mode)
-{
-	struct tm *tm;
-	static struct strbuf timebuf = STRBUF_INIT;
-
-	if (mode == DATE_RAW) {
-		strbuf_reset(&timebuf);
-		strbuf_addf(&timebuf, "%lu %+05d", time, tz);
-		return timebuf.buf;
-	}
-
-	if (mode == DATE_RELATIVE) {
-		struct timeval now;
-
-		strbuf_reset(&timebuf);
-		gettimeofday(&now, NULL);
-		show_date_relative(time, tz, &now, &timebuf);
-		return timebuf.buf;
-	}
-
-	if (mode == DATE_LOCAL)
-		tz = local_tzoffset(time);
-
-	tm = time_to_tm(time, tz);
-	if (!tm) {
-		tm = time_to_tm(0, 0);
-		tz = 0;
-	}
-
-	strbuf_reset(&timebuf);
-	if (mode == DATE_SHORT)
-		strbuf_addf(&timebuf, "%04d-%02d-%02d", tm->tm_year + 1900,
-				tm->tm_mon + 1, tm->tm_mday);
-	else if (mode == DATE_ISO8601)
-		strbuf_addf(&timebuf, "%04d-%02d-%02d %02d:%02d:%02d %+05d",
-				tm->tm_year + 1900,
-				tm->tm_mon + 1,
-				tm->tm_mday,
-				tm->tm_hour, tm->tm_min, tm->tm_sec,
-				tz);
-	else if (mode == DATE_RFC2822)
-		strbuf_addf(&timebuf, "%.3s, %d %.3s %d %02d:%02d:%02d %+05d",
-			weekday_names[tm->tm_wday], tm->tm_mday,
-			month_names[tm->tm_mon], tm->tm_year + 1900,
-			tm->tm_hour, tm->tm_min, tm->tm_sec, tz);
-	else
-		strbuf_addf(&timebuf, "%.3s %.3s %d %02d:%02d:%02d %d%c%+05d",
-				weekday_names[tm->tm_wday],
-				month_names[tm->tm_mon],
-				tm->tm_mday,
-				tm->tm_hour, tm->tm_min, tm->tm_sec,
-				tm->tm_year + 1900,
-				(mode == DATE_LOCAL) ? 0 : ' ',
-				tz);
-	return timebuf.buf;
-}
-
 /*
  * Check these. And note how it doesn't do the summer-time conversion.
  *
@@ -763,7 +644,7 @@
 	else if (!strcmp(format, "raw"))
 		return DATE_RAW;
 	else
-		die("unknown date format %s", format);
+		error("unknown date format %s", format);
 }
 
 void datestamp(char *buf, int bufsize)
@@ -1057,15 +938,18 @@
 	int number = 0;
 	int touched = 0;
 	struct tm tm, now;
-	time_t time_sec;
+	time_t time_sec, n;
+	int fill_mode = 1; 	/* 1: current time, 0: zero */
 
 	time_sec = tv->tv_sec;
-	localtime_r(&time_sec, &tm);
-	now = tm;
+	localtime_r(&time_sec, &now);
 
 	tm.tm_year = -1;
 	tm.tm_mon = -1;
 	tm.tm_mday = -1;
+	tm.tm_hour = -1;
+	tm.tm_min = -1;
+	tm.tm_sec = -1;
 
 	for (;;) {
 		unsigned char c = *date;
@@ -1084,7 +968,41 @@
 	pending_number(&tm, &number);
 	if (!touched)
 		*error_ret = 1;
-	return update_tm(&tm, &now, 0);
+
+	/* fill in missing members */
+	if (tm.tm_year < 0) {
+	        tm.tm_year = fill_mode ? now.tm_year : 0;
+	} else {
+ 	        fill_mode = 0;
+	}
+	if (tm.tm_mon < 0) {
+	        tm.tm_mon = fill_mode ? now.tm_mon : 0;
+	} else {
+	        fill_mode = 0;
+	}
+	if (tm.tm_mday < 0) {
+   	        tm.tm_mday = fill_mode ? now.tm_mday : 1;
+	} else {
+	        fill_mode = 0;
+	}
+	if (tm.tm_hour < 0) {
+	        tm.tm_hour = fill_mode ? now.tm_hour : 0;
+	} else {
+	        fill_mode = 0;
+	}
+	if (tm.tm_min < 0) {
+	        tm.tm_min = fill_mode ? now.tm_min : 0;
+	} else {
+	        fill_mode = 0;
+	}
+	if (tm.tm_sec < 0) {
+	        tm.tm_sec = fill_mode ? now.tm_sec : 0;
+	} else {
+	        fill_mode = 0;
+	}
+
+	n = timegm(&tm);
+	return n;
 }
 
 unsigned long approxidate_relative(const char *date, const struct timeval *tv)
